db2i_file.cc

```diff
@@ -387,9 +387,10 @@ void db2i_table::filenameToTablename(const char* in, char* out, size_t outlen)
     strcat(out, "#P#");
     accumLen += 4;
     
-    memset(temp, 0, min(outlen, part2-part1));
-    memcpy(temp, part3, min(outlen, part4-part3));
-    temp[min(outlen-1, part4-part3)] = 0;
+    memset(temp, 0, std::min(outlen, num_bytes));
+    size_t num_bytes2 = std::min(outlen, static_cast<size_t>(part4-part3));
+    memcpy(temp, part3, std::min(outlen, num_bytes2));
+    temp[std::min(static_cast<size_t>(outlen-1), num_bytes2)] = 0;
```

Comment:
Same as above. Also, we never use num_bytes from above again, so we could re-use it here instead of making num_bytes2.

Might be worth wrapping this code in to a function / lambda which can be called. Essentially what's being done is similar to strlcpy, eg.

```c++
// like strclpy, but takes srclen parameter too
auto strlcpy2 = [](char* dst, size_t size, const char* src, size_t srclen) {
  size = std::min(size-1, srclen);
  memcpy(dst, src, size);
  dst[size] = 0;
  return srclen;
}
```

That way we don't need the num_bytes variable around at all and the callsite is simplified to:

```c++
strlcpy2(tmp, outlen, part3, part4-part3);
```

Of course, as I look further, the only reason we are doing this copy in the first place is to extract a substring from `in` that is null-terminated so we can pass that string to `smartFilenameToTableName`. If `smartFilenameToTableName` took a "source length" parameter (instead of just assuming null-termination), we wouldn't even need to make any copies at all and most of this code would go away.

Maybe something like:

```c++

size_t db2i_table::smartFilenameToTableName(const char *in, size_t inlen, char* out, size_t outlen) {
 // copy old code here and adjust, using inlen
}

size_t db2i_table::smartFilenameToTableName(const char *in, char* out, size_t outlen) {
  return smartFilenameToTableName(in, strlen(in), out, outlen);
}
...

accumLen += smartFilenameToTableName(part3, part4-part3, &out[accumLen], outlen-accumLen);

```

Of course, now I'm really off in to the weeds for this change.

Reply:
We talked about this yesterday, the simplest "fix" for this is to just wrap these in a new scope:
```c++
{
  size_t num_bytes = std::min(outlen, static_cast<size_t>(part2 - part1));
  memcpy(temp, part1, std::min(outlen, num_bytes));
  temp[std::min(outlen-1, num_bytes)] = 0;
}
```

This way num_bytes cannot be used outside the scope and we can re-use the num_bytes name for each without contents of the variable accidentally affecting something else.

Reply:
Updated